### PR TITLE
New fluent API

### PR DESCRIPTION
**Pull request by WojciechNagorski**
Fixes for #1234 
The most significant changes:

- **Changes in Job configuration**

NewApi

- **Changes in IConfig/ManualConfig**

NewApi2

- **Full fluent API**

NewApi3

### CODE REVIEW

File: src/BenchmarkDotNet/Configs/ConfigExtensions.cs [hunk 1]
```diff
@@ -18,38 +18,90 @@ namespace BenchmarkDotNet.Configs
...
+        [Obsolete("This property will soon be removed, please start using WithOptions instead.")]
+        [PublicAPI] public static IConfig With(this IConfig config, ConfigOptions options) => WithOptions(config, options);
+        [PublicAPI] public static ManualConfig WithOptions(this IConfig config, ConfigOptions options) => config.With(m => m.WithOptions(options));
+        [PublicAPI] public static ManualConfig WithoutOptions(this IConfig config, ConfigOptions options) => config.With(m => m.WithoutOptions(options));
+        [PublicAPI] public static ManualConfig WithOptionsIf(this IConfig config, bool value, ConfigOptions options) => config.With(m => m.WithOptionsIf(value, options));
+        
+
+        [Obsolete("This property will soon be removed, please start using GroupBenchmarksBy instead.")]
+        [PublicAPI] public static IConfig With(this IConfig config, params BenchmarkLogicalGroupRule[] rules) => GroupBenchmarksBy(config, rules);
+        [PublicAPI] public static ManualConfig GroupBenchmarksBy(this IConfig config, params BenchmarkLogicalGroupRule[] rules) => config.With(c => c.GroupBenchmarksBy(rules));
```
Comment (AndreyAkinshin) [inline src/BenchmarkDotNet/Configs/ConfigExtensions.cs]: I think we should use the `Add*` notation here. E.g., it can be renamed to `AddLogicalGroupRule`.

**Comment by NickCraver** [comment]
I completely agree the overloads are too numerous on `.With()` these days - it's hard to use, but IMO a ton of `.UseX()` methods is swinging too far in the other direction and exchanges one usability issue for another (larger) issue. Practically speaking, I think the PR for this goes _too_ far and swamps the IntelliSense completion list. Let's take 1 example:
```diff
-public static Job With(this Job job, Jit jit) => job.WithCore(j => j.Environment.Jit = jit);
-public static Job With(this Job job, Runtime runtime) => job.WithCore(j => j.Environment.Runtime = runtime);
+[Obsolete("This property will soon be removed, please start using WithDefaultJit, WithLegacyJit, WithRyuJit, WithLlvm or WithJit instead")]
+public static Job With(this Job job, Jit jit) => job.WithJit(jit);
+public static Job WithJit(this Job job, Jit jit) => job.WithCore(j => j.Environment.Jit = jit);
+public static Job WithDefaultJit(this Job job) => job.WithCore(j => j.Environment.Jit = Jit.Default);
+public static Job WithLegacyJit(this Job job) => job.WithCore(j => j.Environment.Jit = Jit.LegacyJit);
+public static Job WithRyuJit(this Job job) => job.WithCore(j => j.Environment.Jit = Jit.RyuJit);
+public static Job WithLlvm(this Job job) => job.WithCore(j => j.Environment.Jit = Jit.Llvm);
```

IMO, `.WithJit()` is a great sweet spot. It let's me (via IntelliSense) see the categories and get to them quickly. Inside that my options are limited and it's easy to complete what I'm after, see the options, etc. However, the additional overloads of `WithDefaultJit`, `WithLegacyJit`, `WithRyuJit`, `WithLlvm` make the problem far worse and cancel out the wins. I think this change would be a lot better _not_ adding those specific ones, which also increase maintenance and Intellisense length/pain every time a new enum value is added.

File: src/BenchmarkDotNet/Configs/ConfigExtensions.cs [hunk 2]
```diff
@@ -18,38 +18,90 @@ namespace BenchmarkDotNet.Configs
...
+        [PublicAPI] public static IConfig With(this IConfig config, params IValidator[] validators) => AddValidator(config, validators);
+        [PublicAPI] public static ManualConfig AddValidator(this IConfig config, params IValidator[] validators) => config.With(m => m.AddValidator(validators));
+
+        [Obsolete("This property will soon be removed, please start using AddJob instead.")]
+        [PublicAPI] public static IConfig With(this IConfig config, params Job[] jobs) => AddJob(config, jobs);
+        [PublicAPI] public static ManualConfig AddJob(this IConfig config, params Job[] jobs) => config.With(m => m.AddJob(jobs));
+
+        [Obsolete("This property will soon be removed, please start using WithOrderer instead.")]
+        [PublicAPI] public static IConfig With(this IConfig config, IOrderer orderer) => WithOrderer(config, orderer);
+        [PublicAPI] public static ManualConfig WithOrderer(this IConfig config, IOrderer orderer) => config.With(m => m.WithOrderer(orderer));
```
Comment (AndreyAkinshin) [inline src/BenchmarkDotNet/Configs/ConfigExtensions.cs]: I think it should be `AddOrderer`.

File: src/BenchmarkDotNet/Configs/ConfigExtensions.cs [hunk 3]
```diff
@@ -18,38 +18,90 @@ namespace BenchmarkDotNet.Configs
...
         /// <summary>
         /// sets given options for the config
         /// </summary>
-        [PublicAPI] public static IConfig With(this IConfig config, ConfigOptions options) => config.With(m => m.Options = options);
-        [PublicAPI] public static IConfig With(this IConfig config, params BenchmarkLogicalGroupRule[] rules) => config.With(c => c.Add(rules));
+        [Obsolete("This property will soon be removed, please start using WithOptions instead.")]
+        [PublicAPI] public static IConfig With(this IConfig config, ConfigOptions options) => WithOptions(config, options);
+        [PublicAPI] public static ManualConfig WithOptions(this IConfig config, ConfigOptions options) => config.With(m => m.WithOptions(options));
+        [PublicAPI] public static ManualConfig WithoutOptions(this IConfig config, ConfigOptions options) => config.With(m => m.WithoutOptions(options));
+        [PublicAPI] public static ManualConfig WithOptionsIf(this IConfig config, bool value, ConfigOptions options) => config.With(m => m.WithOptionsIf(value, options));
```
Comment (AndreyAkinshin) [inline src/BenchmarkDotNet/Configs/ConfigExtensions.cs]: `WithOptionsIf` sounds pretty confusing to me.

Comment (wojtpl2) [reply src/BenchmarkDotNet/Configs/ConfigExtensions.cs]: @AndreyAkinshin I've changed this name to `WithSwitchingOptions`. What do you think? 
I don't like them both, but I don't have a better idea.

File: src/BenchmarkDotNet/Configs/ConfigExtensions.cs (hunk 1 above)
Comment (wojtpl2) [reply src/BenchmarkDotNet/Configs/ConfigExtensions.cs]: @AndreyAkinshin I've renamed it to `WithLogicalGroupRules`.

**Comment by wojtpl2** [comment]
@NickCraver @AndreyAkinshin I've removed unnecessary overloads (`WithLegacyJit`, `WithRyuJit`, ...) I agree that was not the best idea ;)

**Comment by adamsitnik** [comment]
I've asked @KrzysztofCwalina for an API review, let's see what the author of "Framework Design Guidelines" book and also a BenchmarkDotNet user has to say about our new API ;)

I'll do my part of the review later this week

**Comment by WojciechNagorski** [comment]
@adamsitnik @KrzysztofCwalina Is there any update on this topic?

**Comment by adamsitnik** [comment]
@WojciechNagorski please excuse me for such a delay. I've synced your fork with master. Now I am going to review the PR and push changes (if any) to your fork and merge the PR so I can do some refactor later.

**Comment by WojciechNagorski** [comment]
@adamsitnik  thanks for the information. If you have any questions, just ask.

**Pull request merged**